### PR TITLE
OptaPy Blog Post

### PR DESCRIPTION
**Pull request by Christopher-Chianelli**

### CODE REVIEW

File: content/blog/2021/10/05/OptimizeYourPython.adoc [hunk 1]
```diff
@@ -0,0 +1,438 @@
+= Optimize your Python
+Christopher-Chianelli
+2021-10-05
+:page-interpolate: true
+:jbake-type: post
+:jbake-tags: python, integration
```
Comment (ge0ffrey) [inline content/blog/2021/10/05/OptimizeYourPython.adoc:6]: Can we add a :jbake-social_media_share_image: ?

How about something like [optaplanner icon + python icon = "OptaPy"]?
We can add it under the first paragraph too. I 'll quickly create it and mail it to you - your call if you want to add it.

File: content/blog/2021/10/05/OptimizeYourPython.adoc [hunk 2]
```diff
@@ -0,0 +1,438 @@
+= Optimize your Python
```
Comment (ge0ffrey) [inline content/blog/2021/10/05/OptimizeYourPython.adoc:1]: This title implies that we're optimizing Python code.
Instead, we're optimizing planning problem with Python.

Also, it would be good to have keywords in this title, like "AI" and "constraint solver/solving".

A) How about "A new AI constraint solver for Python: OptaPy"
B) "Want to use OptaPlanner in Python? Now you can!" Nah, I wouldn't target the messaging to "OptaPlanner in Python". We want to attrackt new Python users, not migrate the existing Java users (we are still Java first).
C) "A new Python constraint solver is born: OptaPy" (Burr rule: most important keyword first)

-1 for B. I kinda prefer A), but C) is good too. Other proposals welcome :)

File: content/blog/2021/10/05/OptimizeYourPython.adoc [hunk 3]
```diff
@@ -0,0 +1,438 @@
...
+Christopher-Chianelli
+2021-10-05
+:page-interpolate: true
+:jbake-type: post
+:jbake-tags: python, integration
+
+OptaPlanner can now be used in Python using https://pypi.org/project/optapy[OptaPy]!
+Let's create the school timetabling example in pure Python. Feel free to follow along in the https://mybinder.org/v2/gh/optapy/optapy/main?filepath=index.ipynb[OptaPy Jupyter notebook].
```
Comment (ge0ffrey) [inline content/blog/2021/10/05/OptimizeYourPython.adoc:9]: Ok, the first paragraph has a few goals that it needs to attain:

- Convince python people to take a look at this. So ROI. The words "use OptaPlanner in Python" isn't convincing for python users (they don't know optaplanner).
- Assure them that they can use plain Python. "Pure python" is good already, but I think "plain Python" and mentioning pype might be better.
- The binder message is very important, but can be in the 2th paragraph. It's no longer "motivation" (1th para), it's a "call to action" (2th para).
- Assure Java users reading this that we're not migrating to Python. Prove that by mentioning we support Kotlin too. Or the perf difference?
- Explain OptaPy vs OptaPlanner. Explain its essentially a wrapper.

Let me think of a proposal text and suggest it here.

File: content/blog/2021/10/05/OptimizeYourPython.adoc [hunk 4]
```diff
@@ -0,0 +1,438 @@
...
++
+[source,bash]
+----
+python3 -m pip install optapy
+----
+
+== School timetabling
+
+In school timetabling, we have a list of lessons that we need to assign to timeslot and rooms.
+Additionally, we want some constraints to be followed:
```
Comment (ge0ffrey) [inline content/blog/2021/10/05/OptimizeYourPython.adoc:43]: "to be followed" => passive voice
```suggestion
Additionally, we follow these constraints:
```

Comment (ge0ffrey) [inline content/blog/2021/10/05/OptimizeYourPython.adoc:43]: timeslot: plural

```suggestion
In school timetabling, we need to assign a list of lessons to timeslots and rooms.
Additionally, there are some constraints:
```

File: content/blog/2021/10/05/OptimizeYourPython.adoc [hunk 5]
```diff
@@ -0,0 +1,438 @@
...
+|------------|------------|------------|------------|
+| TUE 13:30: | French     |            | Biology    |
+|            | M. Curie   |            | C. Darwin  |
+|            | 10th grade |            | 9th grade  |
+|------------|------------|------------|------------|
+| TUE 14:30: | Chemistry  | Geography  |            |
+|            | M. Curie   | C. Darwin  |            |
+|            | 9th grade  | 10th grade |            |
+|------------|------------|------------|------------|
+----
```
Comment (ge0ffrey) [inline content/blog/2021/10/05/OptimizeYourPython.adoc:424]: I think this output needs to be at the top, to **show** what we're building before we start building it.

Alternatively - even better I think - would be to Inkscape edit the svg next to this png:
https://github.com/kiegroup/optaplanner/blob/main/optaplanner-docs/src/modules/ROOT/images/quickstart/school-timetabling/schoolTimetablingInputOutput.png
and replace the OptaPlanner icon with the words "OptaPy" and use that at the start of this blog post as well as in the optapy.org README. Then the console output can stay at the end.

File: content/blog/2021/10/05/OptimizeYourPython.adoc [hunk 6]
```diff
@@ -0,0 +1,438 @@
...
+To run the application, put all the code
+into a single Python file and run it:
+[source,bash]
+----
+python3 main.py
+----
+
+== Conclusion
+
+With OptaPy, Python projects can now use OptaPlanner without needing to write any Java code.
```
Comment (ge0ffrey) [inline content/blog/2021/10/05/OptimizeYourPython.adoc:437]:
```suggestion
With https://www.optapy.org[OptaPy], Python developers can now use OptaPlanner in plain Python code (no Java coding needed).
```

File: content/blog/2021/10/05/OptimizeYourPython.adoc [hunk 7]
```diff
@@ -0,0 +1,438 @@
...
+
+The `@planning_entity` decorator registers the class as a planning entity, which allows OptaPlanner to assign its planning variables and for it to be used in constraints.
+The `@planning_variable(variable_type, [value_range_provider_refs...])` decorator registers a method as the getter of a planning variable.
+The getter must be named `get<X>` and the setter must be named `set<X>`.
+The first argument, `variable_type`, tells OptaPlanner what type of values OptaPlanner can assign to this planning variable.
+The second argument, `value_range_provider_refs`, tells OptaPlanner what value ranges (more info on that later) it takes its values from.
+
+== Constraints
+
+Constraints tell OptaPlanner the score of the solution. A score represents the quality of a specific solution. The higher the better. OptaPlanner looks for the best solution, which is the solution with the highest score found in the available time. It might be the optimal solution.
```
Comment (emmurphy1) [inline content/blog/2021/10/05/OptimizeYourPython.adoc:162]:
```suggestion
Constraints define the score calculation, or the fitness function, of a planning problem. Each solution of a planning problem is graded with a score. A score represents the quality of a specific solution. The higher the score the better. OptaPlanner looks for the best solution, which is the solution with the highest score found in the available time. It might or might not be the optimal solution.
```

File: content/blog/2021/10/05/OptimizeYourPython.adoc [hunk 8]
```diff
@@ -0,0 +1,438 @@
...
+                [
+                    Joiners.equal(lambda lesson: lesson.timeslot),
+                    Joiners.equal(lambda lesson: lesson.studentGroup)
+                ]) \
+            .penalize("Student group conflict", HardSoftScore.ONE_HARD)
+----
+
+The `@constraint_provider` decorator allows OptaPlanner to use a function as a constraint provider.
+The function must take exactly one argument; the passed argument is a `ConstraintFactory` used for creating constraints.
+For more information, see https://docs.optaplanner.org/latest/optaplanner-docs/html_single/index.html#constraintStreams[Constraint Streams in the OptaPlanner Docs].
```
Comment (emmurphy1) [inline content/blog/2021/10/05/OptimizeYourPython.adoc:229]:
```suggestion
For more information, see https://docs.optaplanner.org/latest/optaplanner-docs/html_single/index.html#constraintStreams[Constraint Streams] in the OptaPlanner documentation.
```

File: content/blog/2021/10/05/OptimizeYourPython.adoc [hunk 9]
```diff
@@ -0,0 +1,438 @@
...
+
+== Gather the domain objects in a planning solution
+
+A TimeTable wraps all Timeslot, Room, and Lesson instances of a single dataset.
```
Comment (emmurphy1) [inline content/blog/2021/10/05/OptimizeYourPython.adoc:233]:
```suggestion
A TimeTable class wraps all Timeslot, Room, and Lesson instances of a single data set.
```

File: content/blog/2021/10/05/OptimizeYourPython.adoc [hunk 10]
```diff
@@ -0,0 +1,438 @@
...
+Furthermore, because it contains all lessons, each with a specific planning variable state, it is a planning solution and it has a score:
```
Comment (emmurphy1) [inline content/blog/2021/10/05/OptimizeYourPython.adoc:234]:
```suggestion
Furthermore, because it contains all lessons, each with a specific planning variable state, the TimeTable class is a planning solution and it has a score:
```

File: content/blog/2021/10/05/OptimizeYourPython.adoc [hunk 11]
```diff
@@ -0,0 +1,438 @@
...
+    def setScore(self, score):
+        self.score = score
+
+    def __str__(self):
+        return "TimeTable(timeSlotList=" + str(self.timeslotList) + \
+               ",\nroomList=" + str(self.roomList) + ",\nlessonList=" + str(self.lessonList) + \
+               ",\nscore=" + (str(self.score.toString()) if self.score is not None else 'None') + ")"
+----
+
+The `@planning_solution` decorator tells OptaPlanner the class `TimeTable` holds the input and output data.
```
Comment (emmurphy1) [inline content/blog/2021/10/05/OptimizeYourPython.adoc:282]:
```suggestion
The `@planning_solution` decorator tells OptaPlanner that the class `TimeTable` holds the input and output data.
```

File: content/blog/2021/10/05/OptimizeYourPython.adoc [hunk 12]
```diff
@@ -0,0 +1,438 @@
...
+The `@problem_fact_collection_property(fact_type)` decorator tells OptaPlanner the function provide problem facts.
```
Comment (emmurphy1) [inline content/blog/2021/10/05/OptimizeYourPython.adoc:283]:
```suggestion
The `@problem_fact_collection_property(fact_type)` decorator tells OptaPlanner the function that provides problem facts.
```

File: content/blog/2021/10/05/OptimizeYourPython.adoc [hunk 13]
```diff
@@ -0,0 +1,438 @@
...
+The `fact_type` argument tell OptaPlanner what type of fact it provides (for instance, Rooms).
+
+The `@value_range_provider(range_id)` decorator tell OptaPlanner the function provide a value range, which is used to get possible values of planning variables.
```
Comment (emmurphy1) [inline content/blog/2021/10/05/OptimizeYourPython.adoc:286]:
```suggestion
The `@value_range_provider(range_id)` decorator tell OptaPlanner the function that provides a value range, which is used to get possible values of planning variables.
```

File: content/blog/2021/10/05/OptimizeYourPython.adoc [hunk 14]
```diff
@@ -0,0 +1,438 @@
...
+Its argument, `range_id` is a string which is used in `@planning_variable` decorators to link the two (for example, `@planning_variable(Room, ['roomRange'])` is linked to `@value_range_provider('roomRange')`.
+
+The `@planning_entity_collection_property(entity_type)` decorator tells OptaPlanner the function provide planning entities.
```
Comment (emmurphy1) [inline content/blog/2021/10/05/OptimizeYourPython.adoc:289]:
```suggestion
The `@planning_entity_collection_property(entity_type)` decorator tells OptaPlanner the function that provides planning entities.
```

File: content/blog/2021/10/05/OptimizeYourPython.adoc [hunk 15]
```diff
@@ -0,0 +1,438 @@
...
+The `entity_type` argument tell OptaPlanner what type of entities it provides (for instance, Lessons).
+
+The `@planning_score(score_type)` decorator tells OptaPlanner the function returns the score.
```
Comment (emmurphy1) [inline content/blog/2021/10/05/OptimizeYourPython.adoc:292]:
```suggestion
The `@planning_score(score_type)` decorator tells OptaPlanner the function that returns the score.
```

File: content/blog/2021/10/05/OptimizeYourPython.adoc [hunk 16]
```diff
@@ -0,0 +1,438 @@
...
+
+    def __str__(self):
+        return "TimeTable(timeSlotList=" + str(self.timeslotList) + \
+               ",\nroomList=" + str(self.roomList) + ",\nlessonList=" + str(self.lessonList) + \
+               ",\nscore=" + (str(self.score.toString()) if self.score is not None else 'None') + ")"
+----
+
+The `@planning_solution` decorator tells OptaPlanner the class `TimeTable` holds the input and output data.
+The `@problem_fact_collection_property(fact_type)` decorator tells OptaPlanner the function provide problem facts.
+The `fact_type` argument tell OptaPlanner what type of fact it provides (for instance, Rooms).
```
Comment (emmurphy1) [inline content/blog/2021/10/05/OptimizeYourPython.adoc:284]:
```suggestion
The `fact_type` argument tells OptaPlanner what type of fact it provides (for instance, Rooms).
```

File: content/blog/2021/10/05/OptimizeYourPython.adoc [hunk 17]
```diff
@@ -0,0 +1,438 @@
...
+
+The `@value_range_provider(range_id)` decorator tell OptaPlanner the function provide a value range, which is used to get possible values of planning variables.
+Its argument, `range_id` is a string which is used in `@planning_variable` decorators to link the two (for example, `@planning_variable(Room, ['roomRange'])` is linked to `@value_range_provider('roomRange')`.
+
+The `@planning_entity_collection_property(entity_type)` decorator tells OptaPlanner the function provide planning entities.
+The `entity_type` argument tell OptaPlanner what type of entities it provides (for instance, Lessons).
+
+The `@planning_score(score_type)` decorator tells OptaPlanner the function returns the score.
+It must be named `get<X>` and have a corresponding a setter `set<X>`.
+The `score_type` argument tell OptaPlanner what type of score to use (for instance, `HardSoftScore`).
```
Comment (emmurphy1) [inline content/blog/2021/10/05/OptimizeYourPython.adoc:294]:
```suggestion
The `score_type` argument tells OptaPlanner what type of score to use (for instance, `HardSoftScore`).
```

File: content/blog/2021/10/05/OptimizeYourPython.adoc [hunk 18]
```diff
@@ -0,0 +1,438 @@
+= Optimize your Python
+Christopher-Chianelli
+2021-10-05
+:page-interpolate: true
+:jbake-type: post
+:jbake-tags: python, integration
+
+OptaPlanner can now be used in Python using https://pypi.org/project/optapy[OptaPy]!
```
[the remark this answers is not in the document]
Comment (Christopher-Chianelli) [reply content/blog/2021/10/05/OptimizeYourPython.adoc]: https://pypi.org/project/optapy works fine for me

File: content/blog/2021/10/05/OptimizeYourPython.adoc [hunk 19]
```diff
@@ -0,0 +1,438 @@
...
+
+- A student dislikes sequential lessons on the same subject.
+
+== Modelling the domain
+
+The Objects used in constraints are known as the domain of the problem. In school timetabling, this consist of lessons, rooms and timeslots.
+
+=== Problem Facts
+
+Problem facts do not change throughout solving. Rooms and timeslots are examples of problem facts. Let's create the Room class:
```
[the remark this answers is not in the document]
Comment (Christopher-Chianelli) [reply content/blog/2021/10/05/OptimizeYourPython.adoc]: IMO, too wordy, and introduce concepts that are not used (and not currently supported in OptaPy)

File: content/blog/2021/10/05/ANewAIConstraintSolverForPythonOptaPy.adoc [hunk 20]
```diff
@@ -0,0 +1,444 @@
...
+
+- A student dislikes sequential lessons on the same subject.
+
+== Modelling the domain
+
+The Objects used in constraints are known as the domain of the problem. In school timetabling, the domain consist of lessons, rooms and timeslots.
+
+=== Problem Facts
+
+Problem facts do not change throughout solving. Rooms and timeslots are examples of problem facts. Let's create the Room class:
```
Comment (yurloc) [inline content/blog/2021/10/05/ANewAIConstraintSolverForPythonOptaPy.adoc:70]:
```suggestion
Problem facts do not change throughout solving. Rooms and timeslots are examples of problem facts. Let's create the `Room` class:
```
for consistency with the later paragraphs

File: content/blog/2021/10/05/ANewAIConstraintSolverForPythonOptaPy.adoc [hunk 21]
```diff
@@ -0,0 +1,444 @@
...
+        return "TimeTable(timeSlotList=" + str(self.timeslotList) + \
+               ",\nroomList=" + str(self.roomList) + ",\nlessonList=" + str(self.lessonList) + \
+               ",\nscore=" + (str(self.score.toString()) if self.score is not None else 'None') + ")"
+----
+
+The `@planning_solution` decorator tells OptaPlanner that the class `TimeTable` holds the input and output data.
+The `@problem_fact_collection_property(fact_type)` decorator tells OptaPlanner the function that provides problem facts.
+The `fact_type` argument tells OptaPlanner what type of fact it provides (for instance, Rooms).
+
+The `@value_range_provider(range_id)` decorator tell OptaPlanner the function that provides a value range, which is used to get possible values of planning variables.
```
Comment (yurloc) [inline content/blog/2021/10/05/ANewAIConstraintSolverForPythonOptaPy.adoc:293]:
```suggestion
The `@value_range_provider(range_id)` decorator tells OptaPlanner the function that provides a value range, which is used to get possible values of planning variables.
```

File: content/blog/2021/10/05/ANewAIConstraintSolverForPythonOptaPy.adoc [hunk 22]
```diff
@@ -0,0 +1,444 @@
...
+Its argument, `range_id` is a string which is used in `@planning_variable` decorators to link the two (for example, `@planning_variable(Room, ['roomRange'])` is linked to `@value_range_provider('roomRange')`.
+
+The `@planning_entity_collection_property(entity_type)` decorator tells OptaPlanner the function that provides planning entities.
+The `entity_type` argument tell OptaPlanner what type of entities it provides (for instance, Lessons).
```
Comment (yurloc) [inline content/blog/2021/10/05/ANewAIConstraintSolverForPythonOptaPy.adoc:297]:
```suggestion
The `entity_type` argument tells OptaPlanner what type of entities it provides (for instance, Lessons).
```
